src/library/scala/collection/mutable/HashMap.scala

```diff
@@ -205,7 +203,12 @@ class HashMap[K, V](initialCapacity: Int, loadFactor: Double)
   private[this] def put0(key: K, value: V, getOld: Boolean, hash: Int, idx: Int): Some[V] = {
     table(idx) match {
       case null =>
-        table(idx) = new Node[K, V](key, hash, value, null)
+        var i = idx
+        if (contentSize + 1 >= threshold) {
+          growTable(table.length * 2)
+          i = index(hash)
+        }
+        table(i) = new Node[K, V](key, hash, value, null)
       case old =>
```

Comment:
Growing is based on content size, not table size. You have to do it in this case, too. Otherwise you could add new elements into the same hash bucket indefinitely without growing, thus degrading the performance to that of a linked list.

This means you need to trigger a resize if you make it past the correct hash code in the list. This in turn will invalidate the hash bucket, so you need to recompute the index and perform the scan a second time. This is one of the rare cases where the sorted list gets in the way. Without the sorting you could simply prepend the new element (because you already know it's not in the table) but the sorted insert requires another scan.

It makes the logic a lot more complicated and from a performance PoV the optimistic resize upfront is probably better.

Reply:
OK, I see, thanks.